acme_diags/plot/cartopy/qbo_plot.py

```diff
@@ -0,0 +1,162 @@
+    # dict(axis_range=None, axis_scale=None, data=None, data_label=None, data2=None, data2_label=None, label=None)
+
+
+    # Create new figure axis using dimensions from panel (hard coded)
+    ax = fig.add_axes(panel[n])
+    # Plot either a contourf or line plot
+    if plot_type == 'contourf':
+        if 'data' not in z:
+            raise RuntimeError('Must set z["data"] to use plot_type={}.'.format(plot_type))
+        p1 = ax.contourf(x['data'], y['data'], z['data'], color_levels, cmap=plot_colors)
+        cbar = plt.colorbar(p1, ticks=color_ticks)
+        cbar.ax.tick_params(labelsize=label_size)
+    if plot_type == 'line':
+        if 'data2' not in x or 'data2' not in y:
+            raise RuntimeError('Must set data2 for both x and y to use plot_type={}.'.format(plot_type))
+        elif 'data_label' not in x or 'data2_label' not in x:
+            raise RuntimeError('Must set data_label and data2_label for x to use plot_type={}.'.format(plot_type))
+        p1, = ax.plot(x['data'], y['data'], '-ok')
+        p2, = ax.plot(x['data2'], y['data2'], '--or')
+        plt.grid('on')
+        ax.legend((p1, p2), (x['data_label'], x['data2_label']), loc='upper right', fontsize=label_size)
+    ax.set_title(title, size=label_size, weight='demi')
+    ax.set_xlabel(x['label'], size=label_size)
+    ax.set_ylabel(y['label'], size=label_size)
+    plt.yscale(y['axis_scale'])
+    plt.ylim([y['axis_range'][0], y['axis_range'][1]])
+    plt.yticks(size=label_size)
+    plt.xscale(x['axis_scale'])
+    plt.xlim([x['axis_range'][0], x['axis_range'][1]])
+    plt.xticks(size=label_size)
+    return ax
+
+
+def plot(period_new, parameter, test, ref):
+    label_size = 14
+
+    fig = plt.figure(figsize=(14, 14))
+
+    months = min(ref['qbo'].shape[0], test['qbo'].shape[0])
+    x_test, y_test = np.meshgrid(np.arange(0, months), test['level'])
+    x_ref, y_ref = np.meshgrid(np.arange(0, months), ref['level'])
+    cmap2 = plt.cm.RdBu_r
+    color_levels0 = np.arange(-50, 51, 100. / 20.)
+
+    # Panel 0 (Top Left)
+    x = dict(axis_range=[0, months], axis_scale='linear', data=x_test, label=' ')
+    y = dict(axis_range=[100, 1], axis_scale='log', data=y_test, label='hPa')
+    z = dict(data=test['qbo'].T)
+    ax0 = plot_panel(0, fig, 'contourf', label_size, '{} U 5S-5N'.format(test['name']), x, y, z=z,
+               plot_colors=cmap2, color_levels=color_levels0, color_ticks=[-50, -25, -5, 5, 25, 50])
+    # Panel 1 (Middle Left)
+    x = dict(axis_range=[0, months], axis_scale='linear', data=x_ref, label='month')
+    y = dict(axis_range=[100, 1], axis_scale='log', data=y_ref, label='hPa')
+    z = dict(data=ref['qbo'].T)
+    ax1 = plot_panel(1, fig, 'contourf', label_size, '{} U 5S-5N'.format(ref['name']), x, y, z=z,
+               plot_colors=cmap2, color_levels=color_levels0, color_ticks=[-50, -25, -5, 5, 25, 50])
+    # Panel 2 (Top/Middle Right)
+    # TODO: Are the test/ref values correct for data, data_label, data2, data2_label??
```

Comment:
Are these choices correct? These might have gotten mixed up -- looks like test data gets paired with ref data label and vice versa.

Reply:
Hmmm, can you be more specific? This section of codes looks fine to me..

Reply:
@chengzhuzhang `data` is from the `ref` dictionary but `data_label` is from the `test` dictionary. Similarly, `data2` is from the `test` dictionary but `data2_label` is from the `ref` dictionary. 

If we look at the changes from the first commit (i.e., code as it was before I took over the branch), there's a line https://github.com/E3SM-Project/e3sm_diags/pull/290/commits/e77f0653ed1029143542a71893e9adbb9f5d2598#diff-b00e3b398b0699bf12a262ac6cd22034R329: `plot_panel(2,fig,'line',parameter,label_size,'QBO Amplitude \n (period = 20-40 months)','Amplitude (m/s)','Pressure (hPa)',[0,30],[100,1],'linear','log',amp0[:],level0[:],xdata2=amp1[:],ydata2=level0[:],xdata_label='E3SM',xdata2_label='ERA-I')`.

This means the parameters are as follows: `plot_panel(n=2, fig=fig, plottype='line', parameters=parameter, label_size,title='QBO Amplitude \n (period = 20-40 months)', xlabel='Amplitude (m/s)', ylabel='Pressure (hPa)', xaxisrange=[0,30], yaxisrange=[100,1], xaxisscale='linear', yaxisscale='log', xdata=amp0[:], ydata=level0[:], zdata=None, plotcolors=None, xdata2=amp1[:], ydata2=level0[:], xdata_label='E3SM', xdata2_label='ERA-I', color_levels=None, color_ticks=None)`.

That is, `xdata=amp0[:], xdata_label='E3SM', xdata2=amp1[:], xdata2_label='ERA-I'`. `amp0` is now called `ref['amplitude']` and `amp1` is now called `test['amplitude']`. However, `E3SM` appears to be the test name and `ERA-I` appears to be the ref name. 

So, do I have something backwards? Should `amp0` actually be the `test['amplitude']` value? Or is `E3SM` actually the ref name? Or is it correct that test label goes with ref data and vice versa?

Reply:
Okay, I only looked codes before line 80. Yes, for ax2 and ax3, it does look like data and data_label got mixed up here.

Reply:
@chengzhuzhang Ok, I changed it so that `data` and `data_label` refer to `ref` while `data2` and `data2_label` refer to `test`.